app/javascript/hyacinth_ui_v1/components/projects/permissions/PermissionsEditor.js

```diff
@@ -0,0 +1,300 @@
+import React, { useState } from 'react';
+import PropTypes from 'prop-types';
+import { useQuery, useMutation } from '@apollo/react-hooks';
+import { Button, Form, Table } from 'react-bootstrap';
+import produce from 'immer';
+import Select from 'react-select';
+import { remove as arrRemove, sortBy as collectionSortBy } from 'lodash';
+import { useHistory } from 'react-router-dom';
+
+import GraphQLErrors from '../../ui/GraphQLErrors';
+import {
+  getProjectPermissionActionsQuery, getProjectPermissionsQuery, updateProjectPermissionsMutation,
+} from '../../../graphql/projects';
+import { getUsersQuery } from '../../../graphql/users';
+import AddButton from '../../ui/buttons/AddButton';
+import RemoveButton from '../../ui/buttons/RemoveButton';
+import CancelButton from '../../ui/forms/buttons/CancelButton';
+
+function PermissionsEditor(props) {
+  const { readOnly } = props;
+  const { projectStringKey } = props;
+  const history = useHistory();
+  const [permissionsChanged, setPermissionsChanged] = useState(false);
+  const [removedUserIds, setRemovedUserIds] = useState(new Set());
+  const [allUsers, setAllUsers] = useState([]);
```

Comment:
The reason I made it part of the state is because I needed a way to re-render when `useQuery`'s `onCompleted` callback runs for `getUsersQuery`, and the allUsers value is populated.  It seems that `usersLoading` gets set to `false` *before* the `onCompleted` callback runs, so that's why I made use of setState to trigger a re-render.

I guess alternately, I could change:
```
if (actionsLoading || permissionsLoading || usersLoading) return (<></>);
```
to
```
if (actionsLoading || permissionsLoading || usersLoading || allUsers == null) return (<></>);
```
But then what would you recommend as the best way to trigger another render call once allUsers is no longer null?

Reply:
I am confused as to why `allUsers = null` needs to be set. What I was thinking was changing the User's query to:
```
const { loading: usersLoading, error: usersError, data: { users: allUsers } = useQuery(getUsersQuery);
```
And then allUsers should be available for the rest of the component.

Reply:
Destructuring assignment didn't work for this case (not sure why), but I went with the other method that you and I talked about.  Just pushed the change.